### ISSUE TITLE
Fixing/tweaking the Lua language server sections to be a bit more informative 

### ISSUE THREAD

**Issue by Zamachi** [neoclide/coc.nvim#3368]
**Is your feature request related to a problem? Please describe.**
Welp, I installed SpaceVim configuration and am using coc as an `autocomplete ` method and a client for the `LSP layer`. 
What I needed was a language server for Lua, so what do I do? (when the SpaceVim documentation fails me at this)
Turn to **CoC docs** of course! The 2 major places that seem to pop-up for configuring(or using preconfigured extensions) the desired LS are the following:

1. The [Language Servers section - Lua](https://github.com/neoclide/coc.nvim/wiki/Language-servers#lua) language (cause that's the one I needed). There I was met with 3 different language servers, **sumneko's** seemed to be the best one so far, but I've decided to try it **last**, after trying the previous 2 **failed**... Namely, I've checked the links provided for the "how to install the language server" type of thing, and I've done it(at least I think I did it _properly_)... After the installation, all that was left (presumably) was to configure `CocConfig ` to actually use the installed language server(s) using the JSON code snippets provided(of course, adapt it to our specific systems). I've used the `:CocLocalConfig` because I wanted to configure this on per-project basis(not globally), and from what I understood in the docs, the **Workspace config takes priority** over all others. Anyways, I tried configuring the option 1. and 3., that didn't work, the servers were reported "not found". I installed the somneko's server as well(using VScode), then I tried adapting the `coc#config` part into the local configuration JSON file, pointing to all the right directories(using absolute paths), nothing, that didn't help, it spew out a list of errors and I didn't know where to copy-paste the provided code, as it doesn't say where it's supposed to be put at? (might want to add that in the description, I've seen a person put it  copy-pasted into a JSON file...), I even tried exporting the executable to the PATH, but that didn't help either, and I'm 100% confident I didn't fail to structure the JSON file, so there must be something I missed...;

2. The [using CoC Extensions page - implemented coc extensions](https://github.com/neoclide/coc.nvim/wiki/Using-coc-extensions#implemented-coc-extensions) section, which lists the [coc-somneko-lua](https://github.com/xiyaowong/coc-sumneko-lua) extension, which is basically plug-and-play of what I tried to accomplish(but failed...3 times...), after installing that, the server was automatically detected, yay hooray, everything works, I've yet to figure out how to specifically tell a project which language server to use, considering using `:CocLocalConfig` (or global for that matter), failed with multiple errors spat out... 

**Describe the solution you'd like**
As for the somneko's Lua server, you might wanna put the same link found in the Coc extensions section in the server configuration section too! 
If not, you might want to expand on "where to put this piece of code that's provided", I tried extrapolating what you wanted it to do, and just put it in the plain JSON file(the paths were correct), but it spat out certain errors, so I eventually gave up, I've been reading incomplete documentation whole day.  As for the other 2 servers (EmmyLua and Alloyed one), I don't know, I downloaded them, built them, installed them, configured the `CocConfig` to point to the server executables using absolute paths, and nothing, Neovim just said the server wasn't found

**Describe alternatives you've considered**
I don't know, I've read both the SpaceVim(for all of the following: autocomplete, lang#lua and lsp layers) and CoC documentation, trying to see whether I've skipped a step in configuring this, but there's so little info provided, it's as if I have to extrapolate data and connect things in my own head about what am I missing, and to be frank, as a beginner in both Neovim, SpaceVim and CoC, I don't think it should take me weeks to set this up, I'm aiming for the minimum viable configuration. 

**Additional context**
I've written enough of lines, I'm afraid to speak at this point.

**Comment by oblitum** [comment]
Sumneko's server is one of the best, if not the best one, and coc-sumneko-lua is the most easy way to use it on Coc's side. The wiki is open for edits, you may just go there and tell coc-sumneko-lua is an alternative extension to the raw LSP configuration bellow. The wiki page about server configurations exists for showing how to configure servers without the necessity for an extension, so they may or may not be mentioned in there, but they're generally listed in the extensions page, that's why you found it in the extensions page and not on the server configuration page. You should avoid though, telling about SpaceVim specific configuration, etc, because it's off-topic. Personally I use coc-sumneko-lua and never tried any of the other servers.

**Comment by fannheyward** [comment]
> configured the CocConfig to point to the server executables using absolute paths, and nothing, Neovim just said the server wasn't found

Post your coc-settings.json please.

Close this because it's not an issue.

**Comment by Zamachi** [comment]
> Sumneko's server is one of the best, if not the best one, and coc-sumneko-lua is the most easy way to use it on Coc's side. 

Yes, I could tell it's one of the best, hence why I chose it.

> The wiki is open for edits, you may just go there and tell coc-sumneko-lua is an alternative extension to the raw LSP configuration bellow. The wiki page about server configurations exists for showing how to configure servers without the necessity for an extension, so they may or may not be mentioned in there, but they're generally listed in the extensions page, that's why you found it in the extensions page and not on the server configuration page. 

Yes, I could tell it's one of the best, hence why I chose it. I'm afraid to edit anything without anybody's permission even when it says it's open for editing, so yeah, that's why I wrote this suggestion.  Well, I've tried the Wiki way of installign the server via VSCode, then configuring my coc settings(project settings) to point to it, but for some reason that didn't activate...Yeah, they could leave a link like: "...alternatively, you can download [*this*](https://github.com/xiyaowong/coc-sumneko-lua) CoC extension"

> You should avoid though, telling about SpaceVim specific configuration, etc, because it's off-topic. Personally I use coc-sumneko-lua and never tried any of the other servers.

Sorry, didn't know telling about SV is a no-no.

-------------------------------------------------------------------------------------------------------------------------------

> > configured the CocConfig to point to the server executables using absolute paths, and nothing, Neovim just said the server wasn't found
> 
> Post your coc-settings.json please.

Alright, here they are:

```
{
"pyright.enable": true,
"python.pythonPath": "C:/Users/Zamachi/.venvs/cs50-ai/Scripts/python.exe",
"python.venvPath": "C:/Users/Zamachi/.venvs/",
"python.formatting.provider": "yapf",
"python.formatting.yapfPath": "C:/Users/Zamachi/.venvs/cs50-ai/Scripts/yapf.exe",
"python.linting.enabled": true,
"python.linting.pylintEnabled": true,
"python.sortImports.path": "C:/Users/Zamachi/.venvs/cs50-ai/Scripts/isort.exe",
"pyright.organizeimports.provider": "isort"
}

```

Now, I have to say, these are the *last* settings that I've used, I've tried the following *template* before:

```
{
  "languageserver": {
    "go": {
      "command": "gopls",
      "rootPatterns": ["go.mod"],
      "trace.server": "verbose",
      "filetypes": ["go"]
    }
  }
}

```
Where I've assumed "go" is the name of the language, so I'd replace it with *lua* for example, then I think I tried both "command" option(I've had the server on my PATH) and "module" to point to the index.js file of the Somneko server, but neither of them worked and of course, I've specified the filetypes properly...It just seemed that whatever I've typed, the Coc *wouldn't* load the *LocalConfig*, and I didn't have a clue how to check which one is being loaded(if it is being loaded)... From what I've get, I've had it properly set up - there's a .vim dir in my project root with coc-settings.json inside of it.

> 
> Close this because it's not an issue.

Yeah, close it, it isn't an issue, just gave out a suggestion.

**Comment by fannheyward** [comment]
It's a little difficult to run https://github.com/sumneko/lua-language-server with coc.nvim, here's my tests on macOS, you need to change path for Windows.

1. install `sumneko.lua` in VSCode
2. `cd ~/.vscode/extensions/sumneko.lua-2.3.7`
3. `chmod +x ./server/bin/macOS/lua-language-server`
4. `:CocConfig` or `:CocLocalConfig`

```json
{
  "languageserver": {
    "lua": {
      "command": "/Users/fannheyward/.vscode/extensions/sumneko.lua-2.3.7/server/bin/macOS/lua-language-server",
      "args": [
        "-E",
        "/Users/fannheyward/.vscode/extensions/sumneko.lua-2.3.7/server/main.lua"
      ],
      "rootPatterns": [
        ".git"
      ],
      "filetypes": [
        "lua"
      ]
    }
  }
}
```

open a Lua project in vim

**Comment by Zamachi** [comment]
I'm sorry, I've just noticed I've pasted you the Python configuration...which also didn't work for some reason, but the reasoning was the same essentially, proper paths to the Somneko Lua server should've been enough... 

> 3\. `chmod +x ./server/bin/macOS/lua-language-server`, have no idea why this is needed, because in my VSCode extension, the server is not executable.

Wow, I didn't know I need to change the privileges of the server for this? 
I'll try all these steps and see what happens, thanks

> Use https://github.com/xiyaowong/coc-sumneko-lua, it's done the all steps for you, and you don't need to install the VSCode extension because coc-sumneko-lua can download the server for you.

I actually did end up using this method, just installing the extension, which worked without me even providing the coc-settings.json, but that poses the question - what if I want a *different* server for a *different* project, should I just specify this in the `:CocLocaLConfig`?

Thanks for your help!
